src/analysis/normed_space/hahn_banach/separation.lean

```diff
@@ -0,0 +1,353 @@
+  {M₁ : Type*} [add_comm_monoid M₁] {M₂ : Type*} [add_comm_monoid M₂] [module R₁ M₁] [module R₂ M₂]
+  {f : M₁ →ₛₗ[σ₁₂] M₂} (hf : f ≠ 0) :
+  ∃ x, f x ≠ 0 :=
+begin
+  by_contra' h,
+  exact hf (linear_map.ext h),
+end
+
+lemma continuous_linear_map.exists_ne_zero {R₁ R₂ : Type*} [semiring R₁]
+  [semiring R₂] {σ₁₂ : R₁ →+* R₂} {M₁ : Type*} [topological_space M₁] [add_comm_monoid M₁]
+  {M₂ : Type*} [topological_space M₂] [add_comm_monoid M₂] [module R₁ M₁] [module R₂ M₂]
+  {f : M₁ →SL[σ₁₂] M₂} (hf : f ≠ 0) :
+  ∃ x, f x ≠ 0 :=
+begin
+  by_contra' h,
+  exact hf (continuous_linear_map.ext h),
+end
+
+lemma nhds_le_map_nhds [topological_space 𝕜] [topological_space E] {f : E → 𝕜} {g : 𝕜 → E} {a : E}
+  (hg : continuous_at g (f a)) (hcomp : f ∘ g = id) (hgfa : g (f a) = a) :
+  𝓝 (f a) ≤ map f (𝓝 a) :=
+calc 𝓝 (f a) = ((𝓝 (f a)).map g).map f : by rw [map_map, hcomp, map_id]
+  ... ≤ (𝓝 $ g (f a)).map f             : map_mono hg
+  ... = (𝓝 a).map f                     : by rw hgfa
+
+lemma linear_map.nhds_le_map_nhds [topological_space 𝕜] [division_ring 𝕜] [topological_ring 𝕜]
+  [add_comm_group E] [topological_space E] [topological_add_group E] [module 𝕜 E]
+  [has_continuous_smul 𝕜 E] {f : E →ₗ[𝕜] 𝕜} (hf : f ≠ 0) (a : E) :
+  𝓝 (f a) ≤ map f (𝓝 a) :=
+begin
+  obtain ⟨x₀, hx₀⟩ := linear_map.exists_ne_zero hf,
+  let g : 𝕜 → E := λ x, a + (x - f a) • (f x₀)⁻¹ • x₀,
+  have hg : continuous g, by continuity,
+  have hcomp : f ∘ g = id, by { ext, simp [hx₀] },
+  have hgfa : g (f a) = a, by simp [hx₀],
+  exact nhds_le_map_nhds hg.continuous_at hcomp hgfa,
+end
+
+/-- A nonzero continuous linear functional is open. -/
+lemma continuous_linear_map.is_open_map [topological_space 𝕜] [division_ring 𝕜]
+  [topological_ring 𝕜] [add_comm_group E] [topological_space E] [topological_add_group E]
+  [module 𝕜 E] [has_continuous_smul 𝕜 E] (f : E →L[𝕜] 𝕜) (hf : f ≠ 0) :
+  is_open_map f :=
+begin
+  refine is_open_map.of_nhds_le (λ a, _),
+  obtain ⟨x₀, hx₀⟩ := continuous_linear_map.exists_ne_zero hf,
+  let g : 𝕜 → E := λ x, a + (x - f a) • (f x₀)⁻¹ • x₀,
+  have hg : continuous g, by continuity,
+  have hcomp : f ∘ g = id, by { ext, simp [hx₀] },
+  have hgfa : g (f a) = a, by simp [hx₀],
+  exact nhds_le_map_nhds hg.continuous_at hcomp hgfa,
+end
+
+variables [normed_group E]
+
+/-- If `A`, `B` are disjoint sets, `A` is compact and `B` is closed then we can surround them while
+keeping them disjoint. -/
+-- TODO: This proof uses the normed group structure of `E`, but it could work for locally convex
+-- topological vector spaces: instead of taking the balls around 0 with radius 1/n, we could show
+-- there must be some convex neighbourhood `W` of 0 which make `A + W` and `B + W` disjoint?
+theorem exists_disjoint_add_ball {A B : set E} (hA : is_compact A) (hB : is_closed B)
+  (disj : disjoint A B) :
+  ∃ ε : ℝ, 0 < ε ∧ disjoint (A + metric.ball 0 ε) (B + metric.ball 0 ε) :=
```

Comment:
This feels like it should be a result about `metric.thickening`, maybe
```lean
theorem exists_disjoint_thickening {E : Type*} [metric_space E] {A B : set E} (hA : is_compact A)
  (hB : is_closed B)
  (disj : disjoint A B) :
  ∃ ε : ℝ, 0 < ε ∧ disjoint (metric.thickening ε A) (metric.thickening ε B) :=
sorry
```
@kkytola, is this sort of thing available in the API you have been building?

Reply:
I believe it should be relatively straightforward to write `A + metric.ball 0 ε` as the (open) thickening, by either `metric.mem_thickening_iff` or `metric.thickening_eq_bUnion_ball`. Maybe a lemma for specifically this rewrite would be useful.

The `exists_disjoint_thickening` is currently entirely missing from the thickening API. It looks to me it would be quite natural to add, though.

I can try to add these at some point, but if someone wants them earlier, they are certainly welcome. (These could also be left to a later refactor.)

Reply:
I think this sounds like the right sequence of lemmas, indeed.  And I think the lemma is not really satisfactory as is (it's too specific), so I hope Bhavik and Yael could split it off from this PR and do it as part of this project.  I think that there can be a shorter proof of `exists_disjoint_thickening` passing through `metric.continuous_inf_dist_pt B` and `hA.exists_forall_le`.

Reply:
Thanks. This is much needed guidance! I didn't know what `thickening` was about, although I should've guessed.